### ISSUE TITLE
Considering re-visiting our tx-sender strategy

### ISSUE THREAD

**Issue by lgalabru** [blockstack/blockstack-core#982]
Quoting SIP-002:

> Importantly, the tx-sender variable does not change during inter-contract calls. This means that if a transaction invokes a function in a given smart contract, that function is able to make calls into other smart contracts on your behalf. This enables a wide variety of applications, but it comes with some dangers for users of smart contracts. However, as mentioned before, the static analysis guarantees of our smart contracting language allow clients to know a priori which functions a given smart contract will ever call. Good clients should always warn users about any potential side effects of a given transaction.

I'm not sure to understand how "good clients can warn users about any potential side effects of a given transaction", without pre-executing the entire transaction and listing exhaustively the writes (call + values) being performed (probably non trivial in term of engineering + UX).
Even if we end up having clients designing and implementing this requirement, can we guarantee the exact same result with the onchain execution, taking into account the fact that we recently introduced `block-height`? 
The lifecycle of a Tx on our blockchain, from pre-mempool to mining is unclear to me at this point in time, but I'd think that block-height pre-signature and block-height at mining time would probably be different.

With our current model, if, as a user, I decide to trust a contract `XYZ`, then `XYZ` is getting the rights to execute any method on any contract, on my behalf. I feel like I'm giving away my root privileges, without scope.
Let's imagine `XYZ` being malicious / flawed (or is an honest contract relying on a malicious / flawed dependency), and emptying my balances for the 20 most popular tokens + STX.
It sounds like a high price to pay for a user, doing one mistake, one time and trusting the wrong contract.

Am I missing a crucial element in my reasoning?

**Comment by jcnelson** [comment]
> I'm not sure to understand how "good clients can warn users about any potential side effects of a given transaction", without pre-executing the entire transaction and listing exhaustively the writes (call + values) being performed (probably non trivial in term of engineering + UX).

This is _exactly_ the intended behavior -- show the user what the contract might do.  Unlike Solidity, the design of Clarity allows this can be done efficiently and completely.

> Even if we end up having clients designing and implementing this requirement, can we guarantee the exact same result with the onchain execution, taking into account the fact that we recently introduced block-height?

Yes -- because Clarity does not permit dynamic dispatch or late-binding on inter-contract calls, it is tractable to compute the transitive closure of any smart contract's call graph.

> With our current model, if, as a user, I decide to trust a contract XYZ, then XYZ is getting the rights to execute any method on any contract, on my behalf. I feel like I'm giving away my root privileges, without scope.

What you'd be getting is a warning saying that "this transaction _might_ invoke contract XYZ.  Do you wish to proceed?"

**Comment by jcnelson** [comment]
At a higher level, the transitive closure of a transaction can be broad and deep, and the user's policies regarding what contracts to execute (and not execute) will almost certainly be finer-grained than "don't execute this, period."  This is why I opened this issue [here](https://github.com/blockstack/blockstack-core/issues/981), so developers can ensure that a user is only able to invoke their contract directly and not via another smart contract (lest the interposed smart contract do something unintended).

**Comment by lgalabru** [comment]
Instead of spoofing `tx-sender` with `as-contract!`, what about introducing a new intrinsic, `tx-caller`:

- `tx-caller` = `tx-sender` in the root node of the call graph
- `tx-caller` = principal of the calling contract otherwise. 

This would also solve https://github.com/blockstack/blockstack-core/issues/981.

**Comment by jcnelson** [comment]
No, this is undesirable and does not solve the issue.  `tx-sender` needs to resolve to the principal that sent the transaction across contract calls, since this is how we intend to support library contracts.  If I call into contract A, and contract A calls a helper method in contract B, then contract B needs to see the same principal data in `tx-sender` as contract A (since either way I'm the one doing the call).  `as-contract!` is meant to override this behavior _only in situations where this should not be the case_ (which is a small minority of contract calls).

**Comment by lgalabru** [comment]
> No, this is undesirable and does not solve the issue

How having the contract relying on `contract-caller` instead of `tx-sender` is not solving #981 ?

**Comment by jcnelson** [comment]
> How having the contract relying on contract-caller instead of tx-sender is not solving #981 ?

If you have a specific question about the explanation that follows, please ask it specifically.

**Comment by kantai** [comment]
@jcnelson -- I think @lgalabru is asking this question from the background of best practices in Solidity contracts (see, for example, this article: https://vessenes.com/tx-origin-and-ethereum-oh-my/)

Using `tx.origin` (which functions the same as our `tx-sender`) allows contracts to spend tokens to the surprise of the calling user. To some extent, the lack of dynamic dispatch and the ability of the complete call graph to be statically inferable in Clarity solves this problem.

However, I think there may be an argument that even this isn't sufficient. People may want an "allowance" interface (which is the primary purpose of ERC20) rather than a "Prompt if okay before signing" interface. #981 may be a better abstraction for implementing something like this, though I think we should, in that case, integrate that into the naming scheme of functions, so we have "public" functions, which may be called by other smart contracts, and then "transaction" functions, which may only be directly invoked by a user (or, within an `as-contract`?)

**Comment by jcnelson** [comment]
Ah yes, I remember reading that article when it was published.

Reading it again, it makes me feel even more strongly that #981 is necessary.  If I have a wallet contract, then certain functions ought to be privileged such that no contract can invoke them (i.e. the only way to invoke them is through the user directly).  The `msg.sender == owner` check effectively does the same thing, but having a way to define a function this way would make it easy for the developer tooling, the client wallet, and the blockchain VM itself to statically verify that any functions that manipulate tokens or other transferable assets can only be invoked directly by the transaction.

**Comment by jcnelson** [comment]
In fact, we might want to invert this -- maybe instead of explicitly declaring functions as "callable by user only," we instead make functions callable by user only by default.  If a developer wants a function to be callable by a contract in addition to being public, they'll have to explicitly allow it as well.

**Comment by lgalabru** [comment]
I think that #981 is solving one part of the issue I'm trying to highlight.

Let's consider our sample code, and Alice ordering a new `name` on the `names` contract.
```scm
(define-public (preorder 
                (name-hash (buff 20))
                (name-price int))
  (if (is-ok? (contract-call! tokens token-transfer
                burn-address name-price))
      (begin (insert-entry! preorder-map
                     (tuple (name-hash name-hash))
                     (tuple (paid name-price)
                            (buyer tx-sender)))
             (ok 0))
      (err "token payment failed.")))
``` 

When I'm signing a transaction `names preorder 'alice.id 20`, how can I be sure that `names` will only be getting 20 tokens?


If the plan is going for getting an approval from the user, showing this warning upfront:
> This transaction might invoke:
> \- contract `tokens`. 
> Do you wish to proceed?

Is not enough. Instead, we would need to show:
> This transaction might invoke:
> \- contract `tokens`, **transfering 20 tokens to XYZ**.
> Do you wish to proceed?

In other words, we would need to show 
1) the contract call (as planned)
2) the arguments, evaluated.

Is it something we can realistically do, knowing that the evaluation at signature time can be different from the evaluation at mining time?

**Comment by jcnelson** [comment]
> Let's consider block-height being 100 when I signed, and 101 at mining time.
What is blocking me from writing:

Not knowing the block height does not preclude me from efficiently determining the set of reachable code branches, which I can use to tell the user what code will potentially be executed (note that in Solidity, doing this is _not possible_).  This is what I was referring to by "call graph" -- the set of reachable code statements.  Apologies for the confusion.

> Is it something we can realistically do, knowing that the evaluation at pre-signature time can be different from the evaluation at mining time?

No, it is not.  You are correct in saying that the set of code branches executed may be dependent on runtime state that is not known in advance.  For example, the number of tokens transferred can be dependent on the state of the VRF at the given block height -- something that _cannot_ be known in advance.

That said, I don't know what point you're still trying to make?  #981 gives wallet authors a way to prevent contracts from touching user funds, which addresses your concern that `I feel like I'm giving away my root privileges, without scope.`  Also, a user-submitted transaction can already specify pre-conditions and post-conditions of the transaction's evaluation, such that if they are _not_ true, the transaction will abort.  For example, I can have a transaction that tries to preorder a name but aborts if it spends more than 20 tokens -- I simply add a code snippit to check my balance after the `preorder`.  My wallet software can even automatically generate and append this code to my transaction for me.

If you want, we could further refine this concept so the user could specify a whitelist (or blacklist) of contracts and contract methods that can be evaluated at runtime when the transaction is mined, and abort the transaction if a non-whitelisted method is reached.  For example, I could send a transaction that does not whitelist (or explicitly blacklists) `tokens.token-transfer`, so that if the transaction tries to invoke that method, it aborts.

**Comment by kantai** [comment]
Closing in favor of #1051
